File: 0322_coinChange/main.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    int coinChange(std::vector<int>& coins, int amount) {
        std::vector<int> dp(amount + 1, -1);
        dp[0] = 0;
        for(int i = 0; i <= amount; ++i)
        {
            for(int j = 0; j < coins.size(); ++j)
            {
                if(coins[j] <= i && dp[i-coins[j]] != -1 && (dp[i] == -1 || dp[i] > dp[i-coins[j]] + 1))
                    dp[i] = dp[i-coins[j]] + 1;
            }
        }
        return dp[amount];
    }

    int coinChange1(std::vector<int>& coins, int amount) {
        std::vector<int> dp(amount + 1, amount + 1);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int j = 0; j < coins.size(); j++) {
                if (coins[j] <= i) {
                    dp[i] = std::min(dp[i], dp[i - coins[j]] + 1);
                }
            }
        }
        return dp[amount] > amount ? -1 : dp[amount];
    }
};

File: 0322_coinChange/main.py
from typing import List

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount + 1] * (amount + 1)
        dp[0] = 0

        for total in range(1, amount + 1):
            for coin in coins:
                if total - coin >= 0:
                    dp[total] = min(dp[total], 1 + dp[total-coin])

        return dp[amount] if dp[amount] != amount + 1 else -1